Signing support for Miniscript Descriptors

darosior opened the pull request:
This makes the Miniscript descriptors solvable.

Note this introduces signing support for much more complex scripts than the wallet was previously able to solve, and the whole tooling isn't provided for a complete Miniscript integration in the wallet. Particularly, the PSBT<->Miniscript integration isn't entirely covered in this PR.

darosior committed:
miniscript: satisfaction support

This introduces the logic to "sign for" a Miniscript.

Co-Authored-By: Pieter Wuille <pieter.wuille@gmail.com>

darosior committed:
script/sign: basic signing support for Miniscript descriptors

Try to solve a script using the Miniscript satisfier if the legacy
solver fails under P2WSH context. Only solve public key and public key
hash challenges for now.

We don't entirely replace the raw solver and especially rule out trying to
solve CHECKMULTISIG-based multisigs with the Miniscript satisfier since
some features, such as the transaction input combiner, rely on the
specific behaviour of the former.

darosior committed:
script/sign: signing support for Miniscript with timelocks

darosior committed:
script/sign: signing support for Miniscripts with hash preimage challenges

Preimages must be externally provided (typically, via a PSBT).

darosior committed:
qa: functional test Miniscript signing with key and timelocks

We'll need a better integration of the hash preimages PSBT fields to
satisfy Miniscript with such challenges from the RPC.

Thanks to Greg Sanders for his examples and suggestions to improve this
test.

achow101 commented on src/script/miniscript.h:
```diff
@@ -676,6 +744,166 @@ struct Node {
+                }
+                case NodeType::OR_D: {
+                    auto& x = subres[0], &z = subres[1];
+                    auto nsat = z.nsat + x.nsat, sat_l = x.sat, sat_r = z.sat + x.nsat;
```
In b2e3582323b1e3dfded7eb0a2597fd74ba52e4df "miniscript: satisfaction support"

This line appears to not be doing anything.

achow101 replied:
In 3bedfb5b13e863508cb9b0d44fa03cc991151927 "miniscript: satisfaction support"

This line is still here.

darosior replied:
Woops. Done!

achow101 commented on src/test/descriptor_tests.cpp:
```diff
@@ -297,29 +302,39 @@ void DoCheck(const std::string& prv, const std::string& pub, const std::string&
-void Check(const std::string& prv, const std::string& pub, const std::string& norm_prv, const std::string& norm_pub, int flags, const std::vector<std::vector<std::string>>& scripts, const std::optional<OutputType>& type, const std::set<std::vector<uint32_t>>& paths = ONLY_EMPTY)
+void Check(const std::string& prv, const std::string& pub, const std::string& norm_prv, const std::string& norm_pub,
+           int flags, const std::vector<std::vector<std::string>>& scripts, const std::optional<OutputType>& type,
+           const std::set<std::vector<uint32_t>>& paths = ONLY_EMPTY, uint32_t spender_nlocktime=0, uint32_t spender_nsequence=0)
```
In 96e9de9304edf5feec87adf5ee56ad0f526459b8 "script/sign: signing support for Miniscript with timelocks"

In `DoCheck`, `spender_nsequence` is `CTxIn::SEQUENCE_FINAL`, should it be that here as well (instead of 0)?

darosior replied:
Yeah, whichever works. I'm putting `CTxIn::SEQUENCE_FINAL`.

achow101 reviewed (commented):
> Particularly, the PSBT<->Miniscript integration isn't part of this PR.

I don't think it would be particularly hard considering that you're already using `SignatureData` for non-wallet data lookups. I think all that would really need to be done is to update `FillSignatureData` to include the preimages from the PSBT.

Edit: This diff does the trick

```diff
diff --git a/src/psbt.cpp b/src/psbt.cpp
index 8248609ba6..9e09ccf15d 100644
--- a/src/psbt.cpp
+++ b/src/psbt.cpp
@@ -113,6 +113,18 @@ void PSBTInput::FillSignatureData(SignatureData& sigdata) const
     for (const auto& key_pair : hd_keypaths) {
         sigdata.misc_pubkeys.emplace(key_pair.first.GetID(), key_pair);
     }
+    for (const auto& [hash, preimage] : ripemd160_preimages) {
+        sigdata.hash_preimages.emplace(std::vector<unsigned char>(hash.begin(), hash.end()), preimage);
+    }
+    for (const auto& [hash, preimage] : sha256_preimages) {
+        sigdata.hash_preimages.emplace(std::vector<unsigned char>(hash.begin(), hash.end()), preimage);
+    }
+    for (const auto& [hash, preimage] : hash160_preimages) {
+        sigdata.hash_preimages.emplace(std::vector<unsigned char>(hash.begin(), hash.end()), preimage);
+    }
+    for (const auto& [hash, preimage] : hash256_preimages) {
+        sigdata.hash_preimages.emplace(std::vector<unsigned char>(hash.begin(), hash.end()), preimage);
+    }
 }
 
 void PSBTInput::FromSignatureData(const SignatureData& sigdata)
```

darosior committed:
qa: add a fuzz target generating random nodes from a binary encoding

This is a "dumb" way of randomly generating a Miniscript node from
fuzzer input. It defines a strict binary encoding and will always generate
a node defined from the encoding without "helping" to create valid nodes.
It will cut through as soon as it encounters an invalid fragment so
hopefully the fuzzer can tend to learn the encoding and generate valid
nodes with a higher probability.

On a valid generated node a number of invariants are checked, especially
around the satisfactions and testing them against the Script
interpreter.

The node generation and testing is modular in order to later introduce
other ways to generate nodes from fuzzer inputs with minimal code.

Co-Authored-By: Pieter Wuille <pieter@wuille.net>

darosior commented on src/script/miniscript.cpp:
```diff
@@ -282,6 +282,77 @@ size_t ComputeScriptLen(NodeType nodetype, Type sub0typ, size_t subsize, uint32_
+    if (b.available == Availability::NO) return a;
+    // If both are valid, they must be distinct.
+    if (nonmalleable) {
+        // If both options are weak, any result is fine; it just needs the malleable marker.
```
The website defines it as `HASSIG`. That a satisfaction is weak means that it isn't protected by a signature, strong is the opposite. Note this property can be analyzed statically on the Witness Script: it is the 's' property (`NeedsSignature()`, a requirement to be `IsSaneTopLevel()`).

darosior commented on src/script/miniscript.h:
```diff
@@ -225,6 +230,54 @@ size_t ComputeScriptLen(NodeType nodetype, Type sub0typ, size_t subsize, uint32_
+    //! Construct a valid single-element stack (with an element up to 75 bytes).
+    InputStack(std::vector<unsigned char> in) : size(in.size() + 1), stack(Vector(std::move(in))) {}
+    //! Change availability
+    InputStack& Available(Availability avail);
+    //! Mark this input stack as having a signature.
+    InputStack& WithSig();
+    //! Mark this input stack as non-canonical (known to not be necessary in non-malleable satisfactions).
+    InputStack& NonCanon();
+    //! Mark this input stack as malleable.
+    InputStack& Malleable(bool x = true);
```
Done

darosior commented:
Rebased, addressed Andrew's comments, added some cleanups from @sipa, and added a `miniscript_random` fuzz target which generates a random Miniscript node based on a binary encoding. We might eventually add another one which tries to be smart, either here or in a follow-up.

> I don't think it would be particularly hard considering that you're already using SignatureData for non-wallet data lookups. I think all that would really need to be done is to update FillSignatureData to include the preimages from the PSBT.

Yeah, but i figured it would be cleaner to have this as part of a PR which'd also update the various PSBT RPCs and test it using those.

darosior commented:
Fixed the fuzzer crash on integer overflow, and pushed another bunch of updates to the satisfaction code:
- merge the malleable and non-malleable satisfaction
- rename `Choose` to `operator|`
- many improvements to the unit and fuzz tests

See https://github.com/sipa/miniscript/pull/102 for details.

I also pushed an initial corpus for `miniscript_random` fuzz target at https://github.com/bitcoin-core/qa-assets/pull/87

fanquake commented on src/test/descriptor_tests.cpp:
```diff
@@ -342,29 +350,40 @@ void DoCheck(const std::string& prv, const std::string& pub, const std::string&
 
     // Do not replace apostrophes with 'h' in prv and pub
-    DoCheck(prv, pub, norm_prv, norm_pub, flags, scripts, type, paths);
+    DoCheck(prv, pub, norm_prv, norm_pub, flags, scripts, type, paths, /* replace_apostrophe_with_h_in_prv = */false,
```
```suggestion
    DoCheck(prv, pub, norm_prv, norm_pub, flags, scripts, type, paths, /*replace_apostrophe_with_h_in_prv=*/false,
```
Anywhere you are adding / touching named args, please use the [style from the developer docs](https://github.com/bitcoin/bitcoin/blob/master/doc/developer-notes.md#coding-style-c-named-arguments).

darosior committed:
qa: add a "smart" Miniscript fuzz target

At the expense of more complexity, this target generates a valid
Miniscript node at every iteration.

This target will at first run populate a list of recipe (a map from
desired type to possible ways of creating such type) and curate it
(remove the unavailable or redundant recipes).
Then, at each iteration it will pick a type, choose a manner to create a
node of such type from the available recipes, and then
pseudo-recursively do the same for the type constraints of the picked
recipe.

For instance, if it is instructed based on the fuzzer output to create a
Miniscript node of type 'Bd', it could choose to create an 'or_i(subA, subB)'
nodes with type constraints 'B' for subA and 'Bd' for subB. It then
consults the recipes for creating subA and subB, etc...

Here is the list of all the existing recipes, by type constraint:

B: 0()
B: 1()
B: older()
B: after()
B: sha256()
B: hash256()
B: ripemd160()
B: hash160()
B: c:(K)
B: d:(Vz)
B: j:(Bn)
B: n:(B)
B: and_v(V,B)
B: and_b(B,W)
B: or_b(Bd,Wd)
B: or_d(Bdu,B)
B: or_i(B,B)
B: andor(Bdu,B,B)
B: thresh(Bdu)
B: thresh(Bdu,Wdu)
B: thresh(Bdu,Wdu,Wdu)
B: multi()

V: v:(B)
V: and_v(V,V)
V: or_c(Bdu,V)
V: or_i(V,V)
V: andor(Bdu,V,V)

K: pk_k()
K: pk_h()
K: and_v(V,K)
K: or_i(K,K)
K: andor(Bdu,K,K)

W: a:(B)
W: s:(Bo)

Bz: 0()
Bz: 1()
Bz: older()
Bz: after()
Bz: n:(Bz)
Bz: and_v(Vz,Bz)
Bz: or_d(Bzdu,Bz)
Bz: andor(Bzdu,Bz,Bz)
Bz: thresh(Bzdu)

Vz: v:(Bz)
Vz: and_v(Vz,Vz)
Vz: or_c(Bzdu,Vz)
Vz: andor(Bzdu,Vz,Vz)

Bo: sha256()
Bo: hash256()
Bo: ripemd160()
Bo: hash160()
Bo: c:(Ko)
Bo: d:(Vz)
Bo: j:(Bon)
Bo: n:(Bo)
Bo: and_v(Vz,Bo)
Bo: and_v(Vo,Bz)
Bo: or_d(Bodu,Bz)
Bo: or_i(Bz,Bz)
Bo: andor(Bzdu,Bo,Bo)
Bo: andor(Bodu,Bz,Bz)
Bo: thresh(Bodu)

Vo: v:(Bo)
Vo: and_v(Vz,Vo)
Vo: and_v(Vo,Vz)
Vo: or_c(Bodu,Vz)
Vo: or_i(Vz,Vz)
Vo: andor(Bzdu,Vo,Vo)
Vo: andor(Bodu,Vz,Vz)

Ko: pk_k()
Ko: and_v(Vz,Ko)
Ko: andor(Bzdu,Ko,Ko)

Bn: sha256()
Bn: hash256()
Bn: ripemd160()
Bn: hash160()
Bn: c:(Kn)
Bn: d:(Vz)
Bn: j:(Bn)
Bn: n:(Bn)
Bn: and_v(Vz,Bn)
Bn: and_v(Vn,B)
Bn: and_b(Bn,W)
Bn: multi()

Vn: v:(Bn)
Vn: and_v(Vz,Vn)
Vn: and_v(Vn,V)

Kn: pk_k()
Kn: pk_h()
Kn: and_v(Vz,Kn)
Kn: and_v(Vn,K)

Bon: sha256()
Bon: hash256()
Bon: ripemd160()
Bon: hash160()
Bon: c:(Kon)
Bon: d:(Vz)
Bon: j:(Bon)
Bon: n:(Bon)
Bon: and_v(Vz,Bon)
Bon: and_v(Von,Bz)

Von: v:(Bon)
Von: and_v(Vz,Von)
Von: and_v(Von,Vz)

Kon: pk_k()
Kon: and_v(Vz,Kon)

Bd: 0()
Bd: sha256()
Bd: hash256()
Bd: ripemd160()
Bd: hash160()
Bd: c:(Kd)
Bd: d:(Vz)
Bd: j:(Bn)
Bd: n:(Bd)
Bd: and_b(Bd,Wd)
Bd: or_b(Bd,Wd)
Bd: or_d(Bdu,Bd)
Bd: or_i(B,Bd)
Bd: or_i(Bd,B)
Bd: andor(Bdu,B,Bd)
Bd: thresh(Bdu)
Bd: thresh(Bdu,Wdu)
Bd: thresh(Bdu,Wdu,Wdu)
Bd: multi()

Kd: pk_k()
Kd: pk_h()
Kd: or_i(K,Kd)
Kd: or_i(Kd,K)
Kd: andor(Bdu,K,Kd)

Wd: a:(Bd)
Wd: s:(Bod)

Bzd: 0()
Bzd: n:(Bzd)
Bzd: or_d(Bzdu,Bzd)
Bzd: andor(Bzdu,Bz,Bzd)
Bzd: thresh(Bzdu)

Bod: sha256()
Bod: hash256()
Bod: ripemd160()
Bod: hash160()
Bod: c:(Kod)
Bod: d:(Vz)
Bod: j:(Bon)
Bod: n:(Bod)
Bod: or_d(Bodu,Bzd)
Bod: or_i(Bz,Bzd)
Bod: or_i(Bzd,Bz)
Bod: andor(Bzdu,Bo,Bod)
Bod: andor(Bodu,Bz,Bzd)
Bod: thresh(Bodu)

Kod: pk_k()
Kod: andor(Bzdu,Ko,Kod)

Bu: 0()
Bu: 1()
Bu: sha256()
Bu: hash256()
Bu: ripemd160()
Bu: hash160()
Bu: c:(K)
Bu: d:(Vz)
Bu: j:(Bnu)
Bu: n:(B)
Bu: and_v(V,Bu)
Bu: and_b(B,W)
Bu: or_b(Bd,Wd)
Bu: or_d(Bdu,Bu)
Bu: or_i(Bu,Bu)
Bu: andor(Bdu,Bu,Bu)
Bu: thresh(Bdu)
Bu: thresh(Bdu,Wdu)
Bu: thresh(Bdu,Wdu,Wdu)
Bu: multi()

Bzu: 0()
Bzu: 1()
Bzu: n:(Bz)
Bzu: and_v(Vz,Bzu)
Bzu: or_d(Bzdu,Bzu)
Bzu: andor(Bzdu,Bzu,Bzu)
Bzu: thresh(Bzdu)

Bou: sha256()
Bou: hash256()
Bou: ripemd160()
Bou: hash160()
Bou: c:(Ko)
Bou: d:(Vz)
Bou: j:(Bonu)
Bou: n:(Bo)
Bou: and_v(Vz,Bou)
Bou: and_v(Vo,Bzu)
Bou: or_d(Bodu,Bzu)
Bou: or_i(Bzu,Bzu)
Bou: andor(Bzdu,Bou,Bou)
Bou: andor(Bodu,Bzu,Bzu)
Bou: thresh(Bodu)

Bnu: sha256()
Bnu: hash256()
Bnu: ripemd160()
Bnu: hash160()
Bnu: c:(Kn)
Bnu: d:(Vz)
Bnu: j:(Bnu)
Bnu: n:(Bn)
Bnu: and_v(Vz,Bnu)
Bnu: and_v(Vn,Bu)
Bnu: and_b(Bn,W)
Bnu: multi()

Bonu: sha256()
Bonu: hash256()
Bonu: ripemd160()
Bonu: hash160()
Bonu: c:(Kon)
Bonu: d:(Vz)
Bonu: j:(Bonu)
Bonu: n:(Bon)
Bonu: and_v(Vz,Bonu)
Bonu: and_v(Von,Bzu)

Bdu: 0()
Bdu: sha256()
Bdu: hash256()
Bdu: ripemd160()
Bdu: hash160()
Bdu: c:(Kd)
Bdu: d:(Vz)
Bdu: j:(Bnu)
Bdu: n:(Bd)
Bdu: and_b(Bd,Wd)
Bdu: or_b(Bd,Wd)
Bdu: or_d(Bdu,Bdu)
Bdu: or_i(Bu,Bdu)
Bdu: or_i(Bdu,Bu)
Bdu: andor(Bdu,Bu,Bdu)
Bdu: thresh(Bdu)
Bdu: thresh(Bdu,Wdu)
Bdu: thresh(Bdu,Wdu,Wdu)
Bdu: multi()

Wdu: a:(Bdu)
Wdu: s:(Bodu)

Bzdu: 0()
Bzdu: n:(Bzd)
Bzdu: or_d(Bzdu,Bzdu)
Bzdu: andor(Bzdu,Bzu,Bzdu)
Bzdu: thresh(Bzdu)

Bodu: sha256()
Bodu: hash256()
Bodu: ripemd160()
Bodu: hash160()
Bodu: c:(Kod)
Bodu: d:(Vz)
Bodu: j:(Bonu)
Bodu: n:(Bod)
Bodu: or_d(Bodu,Bzdu)
Bodu: or_i(Bzu,Bzdu)
Bodu: or_i(Bzdu,Bzu)
Bodu: andor(Bzdu,Bou,Bodu)
Bodu: andor(Bodu,Bzu,Bzdu)
Bodu: thresh(Bodu)

Co-authored-by: Pieter Wuille <pieter.wuille@gmail.com>

darosior commented:
Rebased, and fixed named args for clang-tidy. This contains the latest iteration of our work on fuzz targets for Miniscript (the `miniscript_stable` and `miniscript_smart` targets introduced in the last two commits, see the commit messages for details).

darosior commented:
Rebased on master and #24148.

darosior commented:
Rebased.

Sjors commented:
> > I don't think it would be particularly hard considering that you're already using SignatureData for non-wallet data lookups. I think all that would really need to be done is to update FillSignatureData to include the preimages from the PSBT.
> 
> Yeah, but i figured it would be cleaner to have this as part of a PR which'd also update the various PSBT RPCs and test it using those.

I created a watch-only wallet for testing miniscript, which relies on the private keys from other wallets. I might be nice to have a (draft) PSBT enable PR on top of this for easier testing.

achow101 commented on src/script/sign.cpp:
```diff
@@ -672,6 +685,7 @@ class DummySignatureChecker final : public BaseSignatureChecker
     bool CheckSchnorrSignature(Span<const unsigned char> sig, Span<const unsigned char> pubkey, SigVersion sigversion, ScriptExecutionData& execdata, ScriptError* serror) const override { return sig.size() != 0; }
     bool CheckLockTime(const CScriptNum& nLockTime) const override { return true; }
     bool CheckSequence(const CScriptNum& nSequence) const override { return true; }
+    bool CheckEqual(const std::vector<unsigned char>& vch1, const std::vector<unsigned char>& vch2) const override { return true; }
```
In 8f1b45bf256b7454d1b790038a00eb537b9bbb70 "script/sign: signing support for Miniscripts with hash preimage challenges"

I think this line should be in the previous commit.

darosior replied:
Why? The previous commit does not need it, it only cares about the interpreter.

achow101 replied:
The previous commit adds `CheckEqual`, whereas this commit adds `LookupPreimage`. It's not immediately obvious how these are related, and IMO it makes more sense to introduce the `CheckEqual`s in the same commit even if unneeded immediately.

darosior replied:
My rationale was that adding `CheckEqual` to the `DummySignatureChecker` is related to the descriptor: it makes a descriptor with timelocks reported as solvable (which in turn allows us to mark them as such in the unit tests). On the other hand the previous commit aims to be the isolated 3-lines-diff that touches consensus code, and only do that.

Not that i feel too strongly about it of course!

darosior commented on src/script/miniscript.h:
```diff
@@ -813,6 +867,189 @@ struct Node {
+                case Fragment::OR_B: {
+                    auto& x = subres[0], &z = subres[1];
+                    // The (sat(Z) sat(X)) solution is overcomplete (attacker can change either into dsat).
+                    return InputResult(z.nsat + x.nsat, (z.nsat + x.sat) | (z.sat + x.nsat) | (z.sat + x.sat).SetMalleable());
```
Indeed. This enables stricter testing. Also, it looks like we are missing a few `SetMalleable()` here, as per the specs:
> The non-canonical options for and_b, or_b, and thresh are always overcomplete (reason 3), so instead use DONTUSE there as well (with HASSIG flag if the original non-canonical solution had one). 

@sipa does the following diff seem correct to you?
```diff
diff --git a/src/script/miniscript.h b/src/script/miniscript.h
index 5c04c40587..e94105d0f4 100644
--- a/src/script/miniscript.h
+++ b/src/script/miniscript.h
@@ -951,12 +951,12 @@ public:
                 }
                 case Fragment::AND_B: {
                     auto& x = subres[0], &y = subres[1];
-                    return InputResult((y.nsat + x.nsat) | (y.sat + x.nsat).SetNonCanon() | (y.nsat + x.sat).SetNonCanon(), y.sat + x.sat);
+                    return InputResult((y.nsat + x.nsat) | (y.sat + x.nsat).SetMalleable().SetNonCanon() | (y.nsat + x.sat).SetMalleable().SetNonCanon(), y.sat + x.sat);
                 }
                 case Fragment::OR_B: {
                     auto& x = subres[0], &z = subres[1];
                     // The (sat(Z) sat(X)) solution is overcomplete (attacker can change either into dsat).
-                    return InputResult(z.nsat + x.nsat, (z.nsat + x.sat) | (z.sat + x.nsat) | (z.sat + x.sat).SetMalleable());
+                    return InputResult(z.nsat + x.nsat, (z.nsat + x.sat) | (z.sat + x.nsat) | (z.sat + x.sat).SetMalleable().SetNonCanon());
                 }
                 case Fragment::OR_C: {
                     auto& x = subres[0], &z = subres[1];
```

sipa replied:
I'd follow @achow101's suggestion but not @darosior's.

The reason is that adding SetMalleable() is "weakening" the self-testing ability, while SetNonCanon() is "strengthening" it. 

The malleability flag influences the signer's choices, so it should only need to be added when we're certain the signer logic can't already figure it out on its own. In OR_B, having an alternative (z.nsat + x.nsat) choice should be sufficient, always available, and not have any signatures. That implies the (z.nsat + x.sat) and (z.sat + x.nsat) choices should be treated as malleable anyway.

On the other hand, the non-canonical flag is effectively an assertion: it doesn't influence the signing logic, except for raising an error if a non-canonical choice ends up being actually used. These should be added wherever we believe, based on higher-level reasoning, that certain choices are off-limits.

darosior replied:
Done.

sipa commented on src/script/interpreter.h:
```diff
@@ -263,6 +263,8 @@ class BaseSignatureChecker
          return false;
     }
 
+    virtual bool CheckEqual(const std::vector<unsigned char>& vch1, const std::vector<unsigned char>& vch2) const { return vch1 == vch2; }
```
I vaguely recall having a discussion about this already, but I can't remember the conclusion.

Is this function needed just to make the dummy signing pass verification in the signing logic? If so, it may be easier to only add a way to bypass that verification check for dummy signing in the first place.

I'm somewhat uncomfortable with making a change to consensus code like this just for that purpose.

darosior replied:
Yes, it's only to pass the `IsSolvable` check and, yes, it is unfortunate. Previous discussion [logs here](https://gnusha.org/miniscript/2022-02-24.log). TL;DR:
- I tested a few things and isolating a 3-lines patch to the interpreter was the only "reasonable" solution i could come up with that didn't change the existing logic
- `IsSolvable` is overkill already, we could just drop the `VerifyScript` there and not need the interpreter patch here:
https://github.com/bitcoin/bitcoin/blob/0897b189e4242321cd03339e7dde5c4d963d0e48/src/script/sign.cpp#L640-L646

But just getting rid of `VerifyScript` isn't enough. We still need to keep `ProduceSignature`, which will in turn call `VerifyScript` again. And since this path is used at "real" signing time too, it gets pretty convoluted to special case `IsSolvable` without using the `DummySignatureChecker`.

I think a cleaner path forward would be to revise the implementation of `IsSolvable` altogether, detangling it from the signing code. It dates from back when there were no descriptors. Nowadays everything we can sign for (that "is solvable") can be represented as a descriptor. So how about re-defining `IsSolvable` as "a descriptor can be inferred from this script, and this descriptor is solvable (ie not raw)"?
It's probably a good cleanup of legacy code on its own too, so it could have its own PR and i would rebase this one on top of it and drop the interpreter patch.

sipa replied:
IIRC, the DummySignatureCreator isn't just used for testing solvability, but also for determining the size of produced signatures before actual keys are available. If so, we still have a dependence on ProduceSignature in that context.

That said, it seems possible to just remove both VerifyScript calls in script/sign.cpp's ProduceSignature and IsSolvable. No tests break.

darosior replied:
Sorry for the late reply.

Yes. I think it would be nice to get rid of it entirely but indeed only redefining `IsSolvable` doesn't permit that, It's sufficient for dropping the interpreter patch here though.

I don't know about removing the belt-and-suspender  `VerifyScript` check in `ProduceSignature`. I'd be more comfortable with just re-defining `IsSolvable`. I've done that in #25664.

darosior committed:
wallet: check solvability using descriptor in AvailableCoins

This is a workaround for Miniscript descriptors containing hash
challenges. For those we can't mock the signature creator without making
OP_EQUAL mockable in the interpreter, so CalculateMaximumInputSize will
always return -1 and outputs for these descriptors would appear
unsolvable while they actually are.

darosior commented:
Rebased. Dropped the commit modifying the interpreter, not needed after #25644. This PR does not touch consensus code anymore.

Note however that since we can't mock the preimage check, we can't estimate the input size when spending from a Miniscript descriptor with a hash preimage challenge in all spending paths. Ideally, i think we should estimate the input size for spending a coin using the descriptors (no need to run the signature producer, we have this information already).
Relatedly, i added a commit for determining solvability using an inferred descriptor in AvailableCoins, as the optimization introduced in 8a105ecd1aeff15f84c3883e2762bf71ad59d920 of using the result of `CalculateMaximumSignedInputSize` would not work for descriptors with hash preimage challenges.

I didn't forget about the draft PR for PSBT support. :) But i'll look first into using descriptors for determining input size.

sipa commented on src/script/sign.h:
```diff
@@ -32,6 +62,9 @@ class BaseSignatureCreator {
     virtual bool CreateSchnorrSig(const SigningProvider& provider, std::vector<unsigned char>& sig, const XOnlyPubKey& pubkey, const uint256* leaf_hash, const uint256* merkle_root, SigVersion sigversion) const =0;
+
+    //! Fetch a preimage corresponding to this hash.
+    virtual bool LookupPreimage(SignatureData& sig_data, const std::vector<unsigned char>& hash, std::vector<unsigned char>& preimage) const =0;
```
I find slightly ugly to expose `SignatureData` to the `SignatureCreator` interface. I'd say that `SignatureData` is the "upper" interface of the script/sign logic, which it exposes to its users. And `SignatureCreator` is part of the "lower" interface - it's how one can provide simple individual signature logic to script/sign. So it feels awkward to make the `SignatureCreator` implementations (which should deal with simple low-level things like signatures) be exposed to this `SignatureData` class which deals with script signing. This isn't more than a philosophical concern about code structure, but I'd like to offer an alternative.

Trying to draw an analogy with how keys/signatures are handled, there is the somewhat weird observation that hash preimages are both key material (which should really go in a `SigningProvider`, as it's a kind of secret) and signature material (which rightfully belongs in `SigningData`, as something that will end up in the witness). I don't think that's a problem - maybe it belongs in both. So a possibility is:
* Add functionality to `SigningProvider` to hold/lookup preimages.
* Have an overridable LookupPreimage in `SigningProvider` that takes a `const SigningProvider&` as argument (like CreateSig etc). It's so simple it could have a default implementation, looking things up in the provider. Dummy implementations could override it by return success with zero-byte ones.
* Still have preimage data in `SignatureData` too, accessed/controlled by a non-overridable ::CreatePreimage function (mimicking ::CreateSig in script/sign.cpp), which can invoke LookupPreimage on the `SignatureCreator` if needed, but can also access the "cached" version in SignatureData.

This means that `SignatureCreator` implementations only need to know about `SigningProvider`, like before, but can still override. It also avoids the need to pass a `SignatureData` to SignSignature.

---

An alternative, if the above is too complex, is to introduce a simple `PreimageData` structure, make `LookupPreimage` take `const PreimageData&`, and add a `PreimageData` field to `SignatureData`.

---

Lastly, if none of the above are doable: can `SignatureData&` here and the argument to SignSignature be made a const reference?

darosior replied:
I see. But since `SignatureData` is already largely exposed to the inner of `ProduceSignature`, what would you think of just making it a parameter in `SignSignature` (instead of creating a fresh one there)? Because actually it's not necessary to expose it to `SignatureCreator`:


<details>
<summary>Quick and dirty diff on the tip of the PR</summary>

```diff
diff --git a/src/script/sign.cpp b/src/script/sign.cpp
index 4fe738dfd5..f337e8ba48 100644
--- a/src/script/sign.cpp
+++ b/src/script/sign.cpp
@@ -89,13 +89,6 @@ bool MutableTransactionSignatureCreator::CreateSchnorrSig(const SigningProvider&
     return true;
 }
 
-bool MutableTransactionSignatureCreator::LookupPreimage(SignatureData& sig_data, const std::vector<unsigned char>& hash, std::vector<unsigned char>& preimage) const {
-    auto it = sig_data.hash_preimages.find(hash);
-    if (it == sig_data.hash_preimages.end()) return false;
-    preimage = it->second;
-    return true;
-}
-
 static bool GetCScript(const SigningProvider& provider, const SignatureData& sigdata, const CScriptID& scriptid, CScript& script)
 {
     if (provider.GetCScript(scriptid, script)) {
@@ -390,6 +383,17 @@ static CScript PushAll(const std::vector<valtype>& values)
     return result;
 }
 
+template<typename M, typename K, typename V>
+bool LookupHelper(const M& map, const K& key, V& value)
+{
+    auto it = map.find(key);
+    if (it != map.end()) {
+        value = it->second;
+        return true;
+    }
+    return false;
+}
+
 /**
  * Context for solving a Miniscript.
  * If enough material (access to keys, hash preimages, ..) is given, produces a valid satisfaction.
@@ -451,16 +455,16 @@ struct Satisfier {
 
     //! Hash preimage satisfactions.
     miniscript::Availability SatSHA256(const std::vector<unsigned char>& hash, std::vector<unsigned char>& preimage) const {
-        return m_creator.LookupPreimage(m_sig_data, hash, preimage) ? miniscript::Availability::YES : miniscript::Availability::NO;
+        return LookupHelper(m_sig_data.hash_preimages, hash, preimage) ? miniscript::Availability::YES : miniscript::Availability::NO;
     }
     miniscript::Availability SatRIPEMD160(const std::vector<unsigned char>& hash, std::vector<unsigned char>& preimage) const {
-        return m_creator.LookupPreimage(m_sig_data, hash, preimage) ? miniscript::Availability::YES : miniscript::Availability::NO;
+        return LookupHelper(m_sig_data.hash_preimages, hash, preimage) ? miniscript::Availability::YES : miniscript::Availability::NO;
     }
     miniscript::Availability SatHASH256(const std::vector<unsigned char>& hash, std::vector<unsigned char>& preimage) const {
-        return m_creator.LookupPreimage(m_sig_data, hash, preimage) ? miniscript::Availability::YES : miniscript::Availability::NO;
+        return LookupHelper(m_sig_data.hash_preimages, hash, preimage) ? miniscript::Availability::YES : miniscript::Availability::NO;
     }
     miniscript::Availability SatHASH160(const std::vector<unsigned char>& hash, std::vector<unsigned char>& preimage) const {
-        return m_creator.LookupPreimage(m_sig_data, hash, preimage) ? miniscript::Availability::YES : miniscript::Availability::NO;
+        return LookupHelper(m_sig_data.hash_preimages, hash, preimage) ? miniscript::Availability::YES : miniscript::Availability::NO;
     }
 };
 
@@ -720,11 +724,6 @@ public:
         sig.assign(64, '\000');
         return true;
     }
-
-    bool LookupPreimage(SignatureData& sig_data, const std::vector<unsigned char>& hash, std::vector<unsigned char>& preimage) const override {
-        preimage.assign(32, '\000');
-        return true;
-    }
 };
 
 }
diff --git a/src/script/sign.h b/src/script/sign.h
index fdb3244d5d..8df1fbc98d 100644
--- a/src/script/sign.h
+++ b/src/script/sign.h
@@ -62,9 +62,6 @@ public:
     /** Create a singular (non-script) signature. */
     virtual bool CreateSig(const SigningProvider& provider, std::vector<unsigned char>& vchSig, const CKeyID& keyid, const CScript& scriptCode, SigVersion sigversion) const =0;
     virtual bool CreateSchnorrSig(const SigningProvider& provider, std::vector<unsigned char>& sig, const XOnlyPubKey& pubkey, const uint256* leaf_hash, const uint256* merkle_root, SigVersion sigversion) const =0;
-
-    //! Fetch a preimage corresponding to this hash.
-    virtual bool LookupPreimage(SignatureData& sig_data, const std::vector<unsigned char>& hash, std::vector<unsigned char>& preimage) const =0;
 };
 
 /** A signature creator for transactions. */
@@ -83,7 +80,6 @@ public:
     const BaseSignatureChecker& Checker() const override { return checker; }
     bool CreateSig(const SigningProvider& provider, std::vector<unsigned char>& vchSig, const CKeyID& keyid, const CScript& scriptCode, SigVersion sigversion) const override;
     bool CreateSchnorrSig(const SigningProvider& provider, std::vector<unsigned char>& sig, const XOnlyPubKey& pubkey, const uint256* leaf_hash, const uint256* merkle_root, SigVersion sigversion) const override;
-    bool LookupPreimage(SignatureData& sig_data, const std::vector<unsigned char>& hash, std::vector<unsigned char>& preimage) const override;
 };
 
 /** A signature creator that just produces 71-byte empty signatures. */
```

</details>

sipa commented on src/script/sign.cpp:
```diff
@@ -443,11 +450,18 @@ struct Satisfier {
+        return m_creator.LookupPreimage(m_sig_data, hash, preimage) ? miniscript::Availability::YES : miniscript::Availability::NO;
+    }
+    miniscript::Availability SatHASH256(const std::vector<unsigned char>& hash, std::vector<unsigned char>& preimage) const {
+        return m_creator.LookupPreimage(m_sig_data, hash, preimage) ? miniscript::Availability::YES : miniscript::Availability::NO;
```
Using a single map/lookup for all types of preimages may be undesirable in an edge case: if `h=SHA256(SHA256(x))`, then both `x` and `SHA256(x)` can be correct preimages (under Hash256 and SHA256 respectively), and both might be used in a single script.

darosior commented:
Rebased (quite verbose, following #25863) and removed needless commit 1af2145ef6 ("descriptor_tests: use clang-tidy verifiable syntax for argument comment").

I changed the commit introducing hashlock satisfaction to use a mapping per hash type, since there might be collisions between hash types (thanks sipa!). I kept a single mapping in the unit tests that is copied in order to not introduce 4 new arguments to the function.
I also changed this commit to not expose `SignatureData` to the `SignatureCreator` anymore. I might change it soon to use a `SigningProvider` to hold the preimages instead, depending on the outcome of the discussion in https://github.com/bitcoin/bitcoin/pull/24149#discussion_r950532827.

darosior commented:
Rebased.

There was a spurious (apparently build?) failure in the Windows CI, had to force-pushed to re-kick it.

darosior commented:
I have implemented in #26567 the signed input size estimation based on descriptor rather than on a dry run of the signing logic. This would allow the wallet to estimate the fees (and therefore do coin selection) for transactions spending Miniscript outputs.

benma commented on src/script/miniscript.cpp:
```diff
@@ -279,6 +279,76 @@ size_t ComputeScriptLen(Fragment fragment, Type sub0typ, size_t subsize, uint32_
+}
+
+InputStack operator|(InputStack a, InputStack b) {
+    // If only one (or neither) is valid, pick the other one.
```
"If only one is invalid, pick the other one. If both are invalid, pick an arbitrary one."

The current comment seems to say the opposite of what the code does.

darosior commented on src/script/sign.cpp:
```diff
@@ -644,19 +662,18 @@ void SignatureData::MergeSignatureData(SignatureData sigdata)
 }
 
-bool SignSignature(const SigningProvider &provider, const CScript& fromPubKey, CMutableTransaction& txTo, unsigned int nIn, const CAmount& amount, int nHashType)
+bool SignSignature(const SigningProvider &provider, const CScript& fromPubKey, CMutableTransaction& txTo, unsigned int nIn, const CAmount& amount, int nHashType, SignatureData sigdata)
```
Oh, i think i initially did that to not have to modify all the call sites (by using a default value in the declaration) when i started writing this and i forgot to change it. Thanks, fixed it (will push once i addressed all comments).

darosior commented on src/test/descriptor_tests.cpp:
```diff
@@ -317,6 +318,11 @@ void DoCheck(const std::string& prv, const std::string& pub, const std::string&
                     txdata.Init(spend, std::move(utxos), /*force=*/true);
                     MutableTransactionSignatureCreator creator{spend, 0, CAmount{0}, &txdata, SIGHASH_DEFAULT};
                     SignatureData sigdata;
+                    // We assume there is no collusion between the hashes (eg h1=SHA256(SHA256(x)) and h2=SHA256(x))
```
Evil hashes colluding!

darosior commented on src/script/sign.cpp:
```diff
@@ -416,9 +484,21 @@ bool ProduceSignature(const SigningProvider& provider, const BaseSignatureCreato
+        // Don't try to satisfy using Miniscript if it was a multisig as it would have already been handled by
+        // SignStep, which appends partial signatures to the witness stack and we must not temper with to keep
+        // compatibility with the extractor relying on this behaviour to combine witnesses.
+        if (!solved && subType != TxoutType::MULTISIG) {
```
Yes, it is equivalent. I've made the change and updated the comment accordingly. Curious why you found `result.clear()` to be slightly-concerning though? It's what we do at the beginning of every `SignStep()` (and the Miniscript satisfier is kind of a more modern `SignStep()`).

darosior replied:
Makes sense.

darosior commented on src/script/miniscript.h:
```diff
@@ -842,6 +1079,47 @@ struct Node {
+                    return std::count(subs.begin(), subs.end(), true) >= node.k;
+                default: // wrappers
+                    assert(subs.size() == 1);
+                    return !!subs[0];
```
Added `-> bool` to the lambda definition as suggested.

darosior commented:
Thanks @sipa for the comment and @benma for having a look. I addressed your comments in the latest push.

achow101 commented:
light ACK 33d30d1edd85a0110cbd9e3e29cc832f1cb6dab2

The satisfaction code looks good to me, as do the fixed tests. I didn't delve too deeply into the fuzz and random tests though.

instagibbs reviewed (commented):
I'd like to boost the tests a bit.

One general case I'd like covered is the one where the PSBT cannot be finalized. Also, I think it'd be useful to assert things like how many partial signatures you expect the wallet to make, which finalization paths it takes depending on nsequence value, things like that.

some examples I found illuminating
```
    # We have one key on each branch; Core signs both (can't finalize)
    {
        "ms": f"c:andor(pk({tprv1}),pk_k({tpubs[0]}),and_v(v:pk({tprv2}),pk_k({tpubs[1]})))",
        "sequence": None,
        "locktime": None,
    },
    # We have all the keys, wallet selects the timeout path to sign since it's smaller and sequence is set
    {
        "ms": f"andor(pk({tprv1}),pk({tprv3}),and_v(v:pk({tprv2}),older(10)))",
        "sequence": 10,
        "locktime": None,
    },
    # We have all the keys, wallet selects the primary path to sign unconditionally since nsequence wasn't set to be valid for timeout path
    {
        "ms": f"andor(pk({tprv1}),pk({tprv3}),and_v(v:pk({tprv2}),older(10)))",
        "sequence": None,
        "locktime": None,
    },
    # Finalizes to the smallest valid witness, regardless of sequence
    {
        "ms": f"or_d(pk({tprv1}),and_v(v:pk({tprv2}),and_v(v:pk({tprv3}),older(10))))",
        "sequence": None,
        "locktime": None,
    },
    # Liquid-like federated pegin with emergency recovery privkeys
    {
        "ms": f"or_i(and_b(pk({tpubs[0]}),a:and_b(pk({tpubs[1]}),a:and_b(pk({tpubs[2]}),a:and_b(pk({tpubs[3]}),s:pk({pk}))))),and_v(v:thresh(2,pkh({tprv1}),a:pkh({tprv2}),a:pkh({tpubs[4]})),older(420)))",
        "sequence": 420,
        "locktime": None,
    },
```

This feature will likely need to be explained in the various PSBT calls as well.

sipa committed:
Align 'e' property of or_d and andor with website spec

sipa committed:
Various additional explanations of the satisfaction logic from Pieter

Cherry-picked and squashed from
https://github.com/sipa/bitcoin/commits/202302_miniscript_improve.

- Explain thresh() and multi() satisfaction algorithms
- Comment on and_v dissatisfaction
- Mark overcomplete thresh() dissats as malleable and explain
- Add comment on unnecessity of Malleable() in and_b dissat

darosior committed:
refactor: make descriptors in Miniscript functional test more readable

We'll add more of them in the next commit, let's keep it bearable.

darosior committed:
psbt: support externally provided preimages for Miniscript satisfaction

Co-Authored-By: Andrew Chow <github@achow101.com>

darosior commented:
@instagibbs thanks for the suggestions and the examples. This functional test was clearly lacking imagination. :) However what you suggest is already thoroughly tested by the unit tests (and to a lesser extent by the fuzz tests). But it's inexpensive to test also in the functional test so i included your examples and added a couple more things to test for each descriptor: the number of signatures expected (always equal to the number of private keys we have in the descriptor), whether this "completes" the transaction or not, and the expected witness stack size of the satisfaction (to assert we used the expected satisfaction path).

@sipa thanks, i think it'll be helpful to reviewers now and to future readers. I squashed [Simplify InputStack constructor calls](https://github.com/sipa/bitcoin/commit/69b1a492c2c2acc70edbf9abe8d433d3d708211f) into the first commit of this PR. I cherry-picked all your commits adding comments and explanations about the satisfaction logic, that i squashed in a single documentation commit (now the second commit of this PR). I was going to say the two commits aligning the properties with the website belong in another PR.. But it's a tiny diff and the `e` property is related to signing, so i just cherry-picked them as well (squashed in one commit).

darosior commented on test/functional/wallet_miniscript.py:
```diff
@@ -20,6 +20,39 @@
+        "sequence": None,
+        "locktime": None,
+    },
+    # A more complex policy, that can't be satisfied through the first branch (need for a preimage)
```
Initially i wanted to tackle all the PSBT integration in a followup (see https://github.com/bitcoin/bitcoin/pull/24149#issuecomment-1046944178). This comment made me (finally) have a look at what should be done. I think `analyzepsbt`, `utxoupdatepsbt` would need to be updated to support hash preimages. Maybe `walletprocesspsbt` too, but i'm not sure yet (for instance hash preimages could be imported separately and kept in a signing provider?). However i think we should first make those commands not rely on a dry run of the signer anymore (Similarly to what i do in https://github.com/bitcoin/bitcoin/pull/26567 with the input size estimation).

Given this followup would be blocked by preparatory work and since the the patch for supporting the preimages provided directly in the PSBT by external software is trivial (basically @achow101's patch from above https://github.com/bitcoin/bitcoin/pull/24149#pullrequestreview-882103227), i added a commit that implements it and tests satisfying using a path with a preimage check.

darosior commented on test/functional/wallet_miniscript.py:
```diff
@@ -20,6 +20,39 @@
+MINISCRIPTS_PRIV = [
+    # One of two keys, of which one private key is known
+    {
+        "ms": "or_i(pk(tprv8ZgxMBicQKsPerQj6m35no46amfKQdjY7AhLnmatHYXs8S4MTgeZYkWAn4edSGwwL3vkSiiGqSZQrmy5D3P5gBoqgvYP2fCUpBwbKTMTAkL/*),pk(tpubD6NzVbkrYhZ4YPAbyf6urxqqnmJF79PzQtyERAmvkSVS9fweCTjxjDh22Z5St9fGb1a5DUCv8G27nYupKP1Ctr1pkamJossoetzws1moNRn/*))",
```
Done. I added a refactoring commit to update the watchonly descriptors, and then use the keys list in the signing test too.

instagibbs commented on test/functional/wallet_miniscript.py:
```diff
@@ -266,6 +286,7 @@ def run_test(self):
                 ms["locktime"],
                 ms["sigs_count"],
                 ms["stack_size"],
+                ms.get("sha256_preimages"),
```
comment above should be updated I think:

"# Test we can sign most Miniscript (all but ones requiring preimages, for now)"

darosior replied:
Thanks, updated.

instagibbs reviewed (commented):
ACK https://github.com/bitcoin/bitcoin/pull/24149/commits/36478ac00ef0a82a82052032e68d513a92ea38b8

reviewed tests only

sipa reviewed (commented):
utACK 6c7a17a8e0eec377f83ed1399f003ae70b898270 (to the extent that it's not my own code).

darosior commented on src/test/orphanage_tests.cpp:
```diff
@@ -108,7 +109,8 @@ BOOST_AUTO_TEST_CASE(DoS_mapOrphans)
         }
-        BOOST_CHECK(SignSignature(keystore, *txPrev, tx, 0, SIGHASH_ALL));
+        SignatureData empty;
+        BOOST_CHECK(SignSignature(keystore, *txPrev, tx, 0, SIGHASH_ALL, empty));
```
Will do if i need to retouch.

achow101 commented:
ACK 6c7a17a8e0eec377f83ed1399f003ae70b898270

Reviewed code and non-fuzz tests.

Sjors commented:
@darosior do I understand correctly that you need #26567 in order for [PSBT stuff](https://github.com/bitcoin/bitcoin/pull/24149#issuecomment-1183382594) to work? Can I just combine both PR's to continue the test I was doing, or are there more changes required?

darosior commented:
@Sjors see https://github.com/bitcoin/bitcoin/pull/24149#discussion_r1103625462. A minimal PSBT integration is part of this PR now. We treat Miniscript-related data when we are given a PSBT. However to fill Miniscript-related data to a PSBT from our wallet will need more work than just #26567, and also some design decisions to be taken.

darosior commented on src/test/fuzz/miniscript.cpp:
```diff
@@ -130,6 +174,732 @@ struct ScriptParserContext {
+using miniscript::operator"" _mst;
+
+//! Construct a miniscript node as a shared_ptr.
+template<typename... Args> NodeRef MakeNodeRef(Args&&... args) { return miniscript::MakeNodeRef<CPubKey>(KEY_COMP, std::forward<Args>(args)...); }
```
https://github.com/bitcoin/bitcoin/pull/27117
